Source/Core/Resource.js

```diff
@@ -0,0 +1,183 @@
+define([
+    './combine',
+    './defaultValue',
+    './defined',
+    './defineProperties',
+    './joinUrls',
+    './objectToQuery',
+    './queryToObject',
+    './Check',
+    '../ThirdParty/Uri'
+], function(combine,
+            defaultValue,
+            defined,
+            defineProperties,
+            joinUrls,
+            objectToQuery,
+            queryToObject,
+            Check,
+            Uri) {
+    'use strict';
+
+    /**
+     * @param {Object} options An object with the following properties
+     * @param {String} options.url
+     * @param {Object} [options.queryParameters]
+     * @param {Object} [options.headers]
+     * @param {Request} [options.request]
+     * @param {String} [options.method='GET']
+     * @param {Object} [options.data]
+     * @param {String} [options.overrideMimeType]
+     * @param {DefaultProxy} [options.proxy]
+     * @param {Boolean} [options.allowCrossOrigin]
+     *
+     * @constructor
+     */
+    function Resource(options) {
+        options = defaultValue(options, defaultValue.EMPTY_OBJECT);
+
+        //>>includeStart('debug', pragmas.debug);
+        Check.defined('options.url', options.url);
+        //>>includeEnd('debug');
+
+        this._url = '';
+        this._queryParameters = defaultValue(options.queryParameters, {});
+        this._hash = defaultValue(options.hash, '');
+
+        this.url = options.url;
+
+        this.headers = options.headers;
+        this.request = options.request;
+        this.responseType = options.responseType;
+        this.method = defaultValue(options.method, 'GET');
+        this.data = options.data;
+        this.overrideMimeType = options.overrideMimeType;
+        this.proxy = options.proxy;
+        this.allowCrossOrigin = defaultValue(options.allowCrossOrigin, true);
+
+        this._retryOnError = options.retryOnError;
+        this._retryAttempts = defaultValue(options.retryAttempts, 0);
+        this._retryCount = 0;
+    }
+
+    defineProperties(Resource.prototype, {
+        url: {
+            get: function() {
+                var uri = new Uri(this._url);
+                uri.query = objectToQuery(this._queryParameters);
+                uri.fragment = this._hash;
+                var url = uri.toString();
+                if (defined(this.proxy)) {
+                    url = this.proxy.getURL(url);
+                }
+                return url;
+            },
+            set: function(value) {
+                var uri = new Uri(value);
+
+                if (defined(uri.query)) {
+                    var query = queryToObject(uri.query);
+                    this._queryParameters = combine(this._queryParameters, query);
+                    uri.query = undefined;
+                }
+
+                if (defined(uri.fragment)) {
```

Comment:
Fragment is an interesting case, since it's not actually sent to the server, so technically it's can't affect a remote resource in any way.  I suppose it makes sense since this encapsulates an URI, though.  I would be curious if any of the existing functionality actually makes use of fragments (I suspect not).